openquake/hazardlib/gsim/mgmpe/generic_gmpe_avgsa.py

```diff
@@ -0,0 +1,198 @@
+        for spec of input and result values.
+        """
+
+        mean_list = []
+        stddvs_list = []
+
+        # Loop over averaging periods
+        for period in self.avg_periods:
+            imt_local = SA(float(period))
+            # compute mean and standard deviation
+            mean, stddvs = self.gmpe.get_mean_and_stddevs(sites, rup, dists,
+                                                          imt_local,
+                                                          stds_types)
+            mean_list.append(mean)
+            stddvs_list.append(stddvs[0]) # Support only for total!
+
+        mean_avgsa = 0.
+        stddvs_avgsa = 0.
+
+        for i1 in range(self.tnum):
+            mean_avgsa += mean_list[i1]
+            for i2 in range(self.tnum):
+                rho = self.corr_func(self.avg_periods[i1],
+                                     self.avg_periods[i2])
+                stddvs_avgsa += rho * stddvs_list[i1] * stddvs_list[i2]
+
+        mean_avgsa *= (1./self.tnum)
+        stddvs_avgsa = (1./self.tnum) * np.sqrt(stddvs_avgsa)
+
+        return mean_avgsa, [stddvs_avgsa]
+
+
+def baker_jayaram_correlation(t1, t2):
+    """
+    Produce inter-period correlation for any two spectral periods.
+    Subroutine taken from: https://usgs.github.io/shakemap/shakelib
+    Based upon:
+    Baker, J.W. and Jayaram, N., 2007, Correlation of spectral acceleration
+    values from NGA ground motion models, Earthquake Spectra.
+
+    :param float t1:
+        First period of interest.
+
+    :param float t2:
+        Second period of interest.
+
+    :return float rho:
+        The predicted correlation coefficient.
+    """
+
+    t_min = min(t1, t2)
+    t_max = max(t1, t2)
+
+    c1 = 1.0 - np.cos(np.pi / 2.0 - np.log(t_max / max(t_min, 0.109)) * 0.366)
+
+    if t_max < 0.2:
+        c2 = 0.105 * (1.0 - 1.0 / (1.0 + np.exp(100.0 * t_max - 5.0)))
+        c2 = 1.0 - c2 * (t_max - t_min) / (t_max - 0.0099)
+    else:
+        c2 = 0
+
+    if t_max < 0.109:
+        c3 = c2
+    else:
+        c3 = c1
+
+    c4 = c1 + 0.5 * (np.sqrt(c3) - c3) * (1.0 + np.cos(np.pi * t_min / 0.109))
+
+    if t_max <= 0.109:
+        rho = c2
+    elif t_min > 0.109:
+        rho = c1
+    elif t_max < 0.2:
+        rho = min(c2, c4)
+    else:
+        rho = c4
+
+    return rho
+
+
+def akkar_correlation(t1, t2):
+    """
+    Read the period-dependent correlation coefficient matrix as in:
+    Akkar S., Sandikkaya MA., Ay BO., 2014, Compatible ground-motion
+    prediction equations for damping scaling factors and vertical to
+    horizontal spectral amplitude ratios for the broader Europe region,
+    Bull Earthquake Eng, 12, pp. 517-547.
+
+    :param float t1:
+        First period of interest.
+
+    :param float t2:
+        Second period of interest.
+
+    :return float:
+        The predicted correlation coefficient.
+    """
+
+    return act.coeff_table[act.periods.index(t1)][act.periods.index(t2)]
```

Comment:
You are not using an interpolation. What's happening when one of the requested periods is not included in the table?

Reply:
Is this a tested case?

Reply:
This is a good point. I would rather not use interpolation, however, but a check on the use of appropriate periods should included, also for PGA. The code presently gives error in that case.

Reply:
Added an error check for periods not list (only Akkar, while baker_jayaram can habdle arbitrary periods). Added also a test for the ValueError.